File: extensions/vscode-aep/src/actions/ActionRegistry.ts
/**
 * Dynamic Action Registry System
 *
 * This provides a plugin-based approach to handling actions from the backend.
 * Instead of hardcoding action types, handlers are registered based on capabilities.
 *
 * Benefits:
 * - Backend can introduce new action types without extension changes
 * - Actions are self-describing based on their structure
 * - Easy to extend with new capabilities
 * - No tight coupling between backend and extension
 */

import * as vscode from 'vscode';

/**
 * Interface for action handlers
 */
export interface ActionHandler {
    /**
     * Unique identifier for this handler
     */
    readonly id: string;

    /**
     * Priority (higher = check first)
     */
    readonly priority: number;

    /**
     * Check if this handler can process the given action
     */
    canHandle(action: any): boolean;

    /**
     * Execute the action
     */
    execute(action: any, context: ActionContext): Promise<ActionResult>;
}

/**
 * Context provided to action handlers
 */
export interface ActionContext {
    workspaceRoot?: string;
    approvedViaChat?: boolean;
    postMessage?: (message: any) => void;
    showMessage?: (message: string) => void;
}

/**
 * Result of action execution
 */
export interface ActionResult {
    success: boolean;
    message?: string;
    data?: any;
    error?: Error;
}

/**
 * Registry for action handlers
 */
export class ActionRegistry {
    private handlers: ActionHandler[] = [];

    /**
     * Register an action handler
     */
    register(handler: ActionHandler): void {
        this.handlers.push(handler);
        // Sort by priority (highest first)
        this.handlers.sort((a, b) => b.priority - a.priority);
        console.log(`[ActionRegistry] Registered handler: ${handler.id} (priority: ${handler.priority})`);
    }

    /**
     * Unregister a handler
     */
    unregister(handlerId: string): void {
        const index = this.handlers.findIndex(h => h.id === handlerId);
        if (index !== -1) {
            this.handlers.splice(index, 1);
            console.log(`[ActionRegistry] Unregistered handler: ${handlerId}`);
        }
    }

    /**
     * Execute an action by finding the appropriate handler
     */
    async execute(action: any, context: ActionContext): Promise<ActionResult> {
        if (!action) {
            return {
                success: false,
                error: new Error('No action provided')
            };
        }

        // Find the first handler that can handle this action
        const handler = this.handlers.find(h => h.canHandle(action));

        if (!handler) {
            console.warn('[ActionRegistry] No handler found for action:', action);
            return {
                success: false,
                error: new Error(`No handler found for action type: ${action.type || 'unknown'}`)
            };
        }

        console.log(`[ActionRegistry] Executing action with handler: ${handler.id}`);

        try {
            return await handler.execute(action, context);
        } catch (error: any) {
            console.error(`[ActionRegistry] Handler ${handler.id} failed:`, error);
            return {
                success: false,
                error: error
            };
        }
    }

    /**
     * Get all registered handlers
     */
    getHandlers(): ReadonlyArray<ActionHandler> {
        return this.handlers;
    }

    /**
     * Get handler by ID
     */
    getHandler(id: string): ActionHandler | undefined {
        return this.handlers.find(h => h.id === id);
    }
}

/**
 * Base class for action handlers
 */
export abstract class BaseActionHandler implements ActionHandler {
    constructor(
        public readonly id: string,
        public readonly priority: number = 50
    ) {}

    abstract canHandle(action: any): boolean;
    abstract execute(action: any, context: ActionContext): Promise<ActionResult>;
}

File: backend/core/action_registry.py
"""
Dynamic Action Registry System for NAVI Backend

This provides a plugin-based approach to handling actions.
Instead of hardcoding action types, handlers are registered based on capabilities.

Benefits:
- Actions can be added without modifying core code
- Handlers are self-describing based on their capabilities
- Easy to extend with new action types
- No tight coupling between intent detection and action execution
"""

from abc import ABC, abstractmethod
from typing import Dict, Any, List, Optional
from dataclasses import dataclass
from enum import Enum
import logging

logger = logging.getLogger(__name__)


class ActionPriority(Enum):
    """Priority levels for action handlers"""

    CRITICAL = 100  # System-level actions
    HIGH = 80  # Core functionality
    NORMAL = 50  # Default priority
    LOW = 20  # Optional/experimental


@dataclass
class ActionContext:
    """Context provided to action handlers"""

    workspace: str
    project_type: str = "unknown"
    current_file: Optional[str] = None
    technologies: List[str] = None
    dependencies: Dict[str, str] = None
    user_id: str = "default"
    session_id: Optional[str] = None

    def __post_init__(self):
        if self.technologies is None:
            self.technologies = []
        if self.dependencies is None:
            self.dependencies = {}


@dataclass
class ActionResult:
    """Result of action execution"""

    success: bool
    action: str
    message: str
    data: Optional[Dict[str, Any]] = None
    files_created: Optional[List[str]] = None
    files_modified: Optional[List[str]] = None
    commands_run: Optional[List[str]] = None
    error: Optional[str] = None
    vscode_command: Optional[Dict[str, Any]] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for API response"""
        result = {
            "success": self.success,
            "action": self.action,
            "message": self.message,
        }

        if self.data:
            result["data"] = self.data
        if self.files_created:
            result["files_created"] = self.files_created
        if self.files_modified:
            result["files_modified"] = self.files_modified
        if self.commands_run:
            result["commands_run"] = self.commands_run
        if self.error:
            result["error"] = self.error
        if self.vscode_command:
            result["vscode_command"] = self.vscode_command

        return result


class ActionHandler(ABC):
    """
    Abstract base class for action handlers

    Handlers register themselves with capabilities and implement
    can_handle() to determine if they can process an action.
    """

    def __init__(
        self, handler_id: str, priority: ActionPriority = ActionPriority.NORMAL
    ):
        self.handler_id = handler_id
        self.priority = priority

    @abstractmethod
    def can_handle(self, action: str, target: str, context: ActionContext) -> bool:
        """
        Determine if this handler can process the given action

        Args:
            action: The action type/name (e.g., "create_component", "open_project")
            target: The action target (e.g., "Button", "my-app")
            context: Execution context with workspace info

        Returns:
            True if this handler can process the action
        """
        pass

    @abstractmethod
    async def execute(
        self, action: str, target: str, context: ActionContext
    ) -> ActionResult:
        """
        Execute the action

        Args:
            action: The action type/name
            target: The action target
            context: Execution context

        Returns:
            ActionResult with success status and details
        """
        pass

    def get_description(self) -> str:
        """Get human-readable description of what this handler does"""
        return f"Handler: {self.handler_id}"

    def get_examples(self) -> List[str]:
        """Get example actions this handler can process"""
        return []


class ActionRegistry:
    """
    Registry for action handlers

    Provides dynamic routing of actions to appropriate handlers
    based on capability matching.
    """

    def __init__(self):
        self.handlers: List[ActionHandler] = []
        self._handler_map: Dict[str, ActionHandler] = {}

    def register(self, handler: ActionHandler) -> None:
        """
        Register an action handler

        Handlers are sorted by priority (highest first)
        """
        self.handlers.append(handler)
        self.handlers.sort(key=lambda h: h.priority.value, reverse=True)

        logger.info(
            f"[ActionRegistry] Registered handler: {handler.handler_id} "
            f"(priority: {handler.priority.name})"
        )

    def unregister(self, handler_id: str) -> None:
        """Unregister a handler by ID"""
        self.handlers = [h for h in self.handlers if h.handler_id != handler_id]
        logger.info(f"[ActionRegistry] Unregistered handler: {handler_id}")

    async def execute(
        self, action: str, target: str, context: ActionContext
    ) -> ActionResult:
        """
        Execute an action by finding the appropriate handler

        Args:
            action: The action type/name
            target: The action target
            context: Execution context

        Returns:
            ActionResult with execution status and details
        """
        if not action:
            return ActionResult(
                success=False,
                action="unknown",
                message="No action specified",
                error="Action is required",
            )

        # Find the first handler that can handle this action
        handler = None
        for h in self.handlers:
            try:
                if h.can_handle(action, target, context):
                    handler = h
                    break
            except Exception as e:
                logger.error(
                    f"[ActionRegistry] Error checking handler {h.handler_id}: {e}"
                )
                continue

        if not handler:
            logger.warning(
                f"[ActionRegistry] No handler found for action: {action} (target: {target})"
            )
            return ActionResult(
                success=False,
                action=action,
                message=f"No handler found for action: {action}",
                error=f"Unsupported action type: {action}",
            )

        logger.info(
            f"[ActionRegistry] Executing action '{action}' with handler: {handler.handler_id}"
        )

        try:
            result = await handler.execute(action, target, context)
            return result
        except Exception as e:
            logger.error(
                f"[ActionRegistry] Handler {handler.handler_id} failed: {e}",
                exc_info=True,
            )
            return ActionResult(
                success=False,
                action=action,
                message=f"Action execution failed: {str(e)}",
                error=str(e),
            )

    def get_handlers(self) -> List[ActionHandler]:
        """Get all registered handlers"""
        return list(self.handlers)

    def get_handler(self, handler_id: str) -> Optional[ActionHandler]:
        """Get handler by ID"""
        return next((h for h in self.handlers if h.handler_id == handler_id), None)

    def get_supported_actions(self) -> List[Dict[str, Any]]:
        """
        Get list of all supported actions from registered handlers

        Returns:
            List of action definitions with name, description, examples
        """
        actions = []
        seen_handlers = set()

        for handler in self.handlers:
            if handler.handler_id in seen_handlers:
                continue

            seen_handlers.add(handler.handler_id)

            actions.append(
                {
                    "handler_id": handler.handler_id,
                    "description": handler.get_description(),
                    "examples": handler.get_examples(),
                    "priority": handler.priority.name,
                }
            )

        return actions


# Global registry instance
_global_registry: Optional[ActionRegistry] = None


def get_action_registry() -> ActionRegistry:
    """Get the global action registry instance"""
    global _global_registry
    if _global_registry is None:
        _global_registry = ActionRegistry()
    return _global_registry


def register_handler(handler: ActionHandler) -> None:
    """Register a handler with the global registry"""
    registry = get_action_registry()
    registry.register(handler)
